Foundation/FileManager.swift

```diff
@@ -442,16 +442,34 @@ open class FileManager : NSObject {
             else {
                 return
         }
+
+        func copyNonDirectory(srcPath: String, dstPath: String, fileType: FileAttributeType) throws {
+            if fileType == .typeSymbolicLink {
+                let destination = try destinationOfSymbolicLink(atPath: srcPath)
+                try createSymbolicLink(atPath: dstPath, withDestinationPath: destination)
+            } else if fileType == .typeRegular {
+                if createFile(atPath: dstPath, contents: contents(atPath: srcPath), attributes: nil) == false {
+                    throw NSError(domain: NSCocoaErrorDomain, code: CocoaError.fileWriteUnknown.rawValue, userInfo: [NSFilePathErrorKey : NSString(string: dstPath)])
+                }
+            }
+        }
+
         if fileType == .typeDirectory {
             try createDirectory(atPath: dstPath, withIntermediateDirectories: false, attributes: nil)
             let subpaths = try subpathsOfDirectory(atPath: srcPath)
             for subpath in subpaths {
-                try copyItem(atPath: srcPath + "/" + subpath, toPath: dstPath + "/" + subpath)
+                let src = srcPath + "/" + subpath
+                let dst = dstPath + "/" + subpath
+                if let attrs = try? attributesOfItem(atPath: src), let fileType = attrs[.type] as? FileAttributeType {
+                    if fileType == .typeDirectory {
+                        try createDirectory(atPath: dst, withIntermediateDirectories: false, attributes: nil)
```

Comment:
This looks wrong. No recursion at all means we only go one extra level deep, no? (There should be a test for this.)

Reply:
The original implementation was wrong because it was recursing in every directory and then additional recursing in every sub directory. It now uses `subpathsOfDirectory(atPath:)` at the top level to get all of the subitems. There is are tests for it at https://github.com/apple/swift-corelibs-foundation/pull/1515/files/3c454b7358312d8d10dcb68a22d577e9ec816b58#diff-fffbc0319b30adac9f635327a750b82fR555

Reply:
Ah, I see. I misread this as `contentsOfDirectory`.

Please note that `subpathsOfDirectory` is very expensive, as it does all the directory enumeration at once and potentially creates a VERY large array. This is a high-water mark issue, similar to the `contentsAtPath` one. Using an enumerator would be much preferable here, and should be a fairly simple drop-in replacement.